### ISSUE TITLE
[QUERY] EventHubs.Message in AppInsights application map

### ISSUE THREAD

**Issue by sebader** [Azure/azure-sdk-for-net#11767]
**Query/Question**
I'm using the Azure.Messaging.EventHubs SDK (version 5.0.1) together with Application Insights SDK (2.13.1). 

````
var batch = await _eventHubProducerClient.CreateBatchAsync(ct);
var data = new EventData(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(myData)));
batch.TryAdd(data);
await _eventHubProducerClient.SendAsync(batch, ct);
````

When sending messages I see this in the Application Insights map:

![image](https://user-images.githubusercontent.com/11445087/80981031-ac13bf00-8e29-11ea-9ddd-f51ba888d525.png)

Details drilldown don't show anything more - and it is not even correlated to the sending operation:
![image](https://user-images.githubusercontent.com/11445087/80981349-25131680-8e2a-11ea-95d6-7526d4c7b740.png)

My question: What does the EventHubs.Message bubble represent and why is it showing up separately from the actual sending operation/dependency?

**Comment by markcowl** [comment]
@sebader Thanks for your question, we will get back to you shortly with a detailed response

**Comment by sebader** [comment]
Any update on this? Thanks!

**Comment by dwoldo** [comment]
I am experiencing the same issue. Does Azure.Messaging.EventHubs support automated telemetry collection? Based on this link: https://docs.microsoft.com/en-us/azure/azure-monitor/app/auto-collect-dependencies it does not.

**Comment by kamilzzz** [comment]
It comes either from EventDataBatch.TryAdd https://github.com/Azure/azure-sdk-for-net/blob/7838c2a8d5058f6e90e369c10aba3b939db74905/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs#L153 or EventProducerClient.SendInternalAsync https://github.com/Azure/azure-sdk-for-net/blob/7838c2a8d5058f6e90e369c10aba3b939db74905/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs#L791 depending if we're sending batch explicitly or IEnumerable of EventBatch.

Anyway, I'm using batched send to EventHub and now in the ApplicationInsights before every `EventHubs.Send` dependency I see 16/32 (depending on my batch sized) `EventHubs.Message` dependencies. 

Any ideas how to disable `EventHubs.Message` dependencies? Obvious way would be to write custom TelemetryProcessor and filter out by dependency name but there are problems with TelemetryProcessors when using Azure Functions v3 with DI.

**Comment by sebader** [comment]
using the latest SDK versions on both producer and consumer, this looks quite nice for me (without any custom code for the instrumentation):
![image](https://user-images.githubusercontent.com/11445087/97545120-d0acb600-19ca-11eb-94b5-213f511657a7.png)

And yes, I'm also seeing the EventHubs.Message but it seems like this serves as the start of the operation for the queue duration calculation.
![image](https://user-images.githubusercontent.com/11445087/97545341-26815e00-19cb-11eb-9dd4-c5d05a166571.png)

The only thing is that the .Send dependency is somehow reported out of band.

**Comment by kamilzzz** [comment]
Ok, so this is looking nice when you have App Insights on both ends (producer & consumer).

For my specific situation I have App Insights connected only on the producer side. Later on it's an 
Azure Stream Analytics or Time Series Insights who's reading from an Event Hub.

That's how it looks like in my case:
![ai](https://user-images.githubusercontent.com/32141632/98179366-df312b00-1efe-11eb-8e99-d0209316dd97.png)

**Comment by alan0609** [comment]
I came across this looking for answers on a similar scenario with the Azuse.Messaging.ServiceBus SDK (7.0.1).  I also see a separate Message entity off from the main ServiceBusSender.Send entity as @sebader is seeing with Event hubs.

**Comment by qetza** [comment]
Hi,
I'm also seeing this where we get a dependency entry for each message created and one for each send. This is not limited to Event Hubs as the Storage SDK also produce multiple dependency entries for a single operation.

The UI stuff is an issue but i see a bigger one: **cost!**
Having two entries for each send means that we double the size of data stored in Application Insights which increase the cost of Application Insights. Currently we are billed daily for 4GB of logs for event hubs dependency entries instead of 2GB ☹️

There should be some easy way to have only a dependency on the send operation. As a workaround we are going to look at filtering some dependencies...

**Comment by lmolkova** [comment]
Thanks for the feedback, it's super useful and we're looking for ways to address it - can you please help and answer a few questions?

@sebader 
Separate Message and Send are expected at this point to make individual messages traceable when they are sent in a batch. What doesn't look right is a separate node for it on the AppMap and OTHER prefix on the Transaction Viewer. On the later screenshots it looks better, do I understand correctly that the visualization issue is resolved, but send/message separation is still an open question? 

@kamilzzz is it fair to assume the duration/success of send operation are the most interesting for you? I.e.
- you're not interested in context propagation
- don't need tracing beyond the producer

@qetza  Are you interested in individual message tracing? I.e.
- do you send batches or individual messages?
- send messages right away without buffering them?
- are you interested in seeing how each individual message are processed on the consumer side?

**Comment by sebader** [comment]
@lmolkova yeah mostly it looks good. One thing I can see is the very last item in the bottom of the screenshot, not sure why Send shows up there:

![image](https://user-images.githubusercontent.com/11445087/135988861-d49ff630-2f52-4cff-b36d-53cd566e1f8c.png)

Application Map is all fine
![image](https://user-images.githubusercontent.com/11445087/135989064-3253a8b8-80f5-4fa8-80af-bbbe342af53b.png)

**Comment by lmolkova** [comment]
@sebader 
> @lmolkova yeah mostly it looks good. One thing I can see is the very last item in the bottom of the screenshot, not sure why Send shows up there:


Thanks for the update! How do you think Send should show up? (assuming you send a batch and have multiple messaged -> processings). 
Our thinking was that perhaps message processing is more important and 'message' -> 'process' shows first. Send is just a transport, assuming it goes well, it's not even interesting, but these are all assumptions and we're happy to hear any feedback.

**Comment by sebader** [comment]
actually I'm not sending batches, just one message per Send operation. It should logically show up before the EventProcessor.Process. From the timeline its not clear why this would show later than Time Spent in Queue?

**Comment by lmolkova** [comment]
I hear your feedback that for the single message it's extra confusing and costly to have a message  + send and then we don't even have a good way to show it.  This is something we should address. I still don't see a way for batching scenarios to be less confusing - happy to hear any ideas or real-life experience.

On a side-note Transaction Viewer is not strictly sorted by timestamp (it gives preference to a parent-child relationship, so process, shows up under message which happens before send), which of  course makes this chart confusing in this case

**Comment by techmantel** [comment]
Facing a similar problem using the `EventHubBufferedProducerClient` in v5.9.2 using OpenTelemetry Azure Monitor exporter v1.0.0-beta.13 and looking at the traces in Application Insights.

I'm having code that listens to MQTT messages and maps them to EventHub messages one by one. For each message I receive from the MQTT broker I start a new `Activity` which is then active when calling `EnqueueEventAsync`. The idea is that I want to batch the messages and send them once every few seconds. 
However the operation id is picked up from the Activity being active during the first message and then that id stays the same for the reminder of the application lifetime (well I've only tested a few minutes anyway).

All messages handled after the first one doesn't show any traces from the EventHub sdk, while the first message contains all of them.

I also tried this code hoping that it would correct it;

```cs
var eventData = new EventData(myMessage.GetBytes())
{
    Properties = {{ "traceparent", Activity.Current?.Id }}
};
```
With this a link in app insights is shown for every message but following that will then populate the UI with all the other messages handled since the start of the application and overloading the UI and making it crash.
![image](https://github.com/Azure/azure-sdk-for-net/assets/4118537/9cd13f55-4f03-4ed7-9c36-3380b1dd965e)

Is this caused for the same reasons or should I open another ticket?
I also found #31922 that looks similar...
